cpu/cortexm_common/periph/systick.c

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (C) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_periph_systick
+ * @{
+ *
+ * @file        systick.c
+ * @brief       Low-level timer driver implementation based on SysTick
+ *
+ *              SysTick is a 24-bit Down-Counting Timer
+ *              It runs at the same frequency as the CPU
+ *              The count (VAL) register can only be set to 0 on write
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include <assert.h>
+#include "board.h"
+#include "cpu.h"
+#include "systick.h"
+
+/* SysTick CTRL configuration */
+#define ENABLE_MASK     (SysTick_CTRL_CLKSOURCE_Msk |\
+                         SysTick_CTRL_ENABLE_Msk    |\
+                         SysTick_CTRL_TICKINT_Msk)
+
+#define ALARM_EXT_MAX   ((1UL << (8 * sizeof(alarm_ext))) - 1)
+
+/* interrupt is always enabled to update time */
+static bool alarm_enabled;
+
+/* divider for CLOCK_CORECLOCK */
+static uint16_t prescaler = 1;
+
+/* full periods before alarm */
+static uint16_t alarm_ext;
+
+/* remaining time to alarm (last period length) */
+static uint32_t alarm_val;
+
+/* current time if the timer ran at freq Hz */
+static uint32_t now;
+
+/* offset in CLOCK_CORECLOCK ticks */
+static uint32_t now_offset;
+
+/* dummy SysTick callback */
+__attribute__((weak))
+void systick_callback(void)
+{
+    /* no-op */
+}
+
+void systick_timer_init(unsigned long freq)
+{
+    if (IS_USED(MODULE_SYSTICK_PRESCALER)) {
+        assert(freq <= CLOCK_CORECLOCK);
+
+        prescaler = CLOCK_CORECLOCK / freq;
+        assert(CLOCK_CORECLOCK % freq == 0);
+    }
+
+    SysTick->VAL  = 0;
+    SysTick->LOAD = SysTick_LOAD_RELOAD_Msk;
+    SysTick->CTRL = ENABLE_MASK;
+
+    NVIC_SetPriority(SysTick_IRQn, CPU_CORTEXM_PENDSV_IRQ_PRIO);
+    NVIC_EnableIRQ(SysTick_IRQn);
+}
+
+int systick_timer_set(unsigned int value)
+{
+    uint64_t val64 = value * prescaler;
+
+    /* check if value fits with prescaler */
+    if ((val64 >> 24) > ALARM_EXT_MAX) {
+        return -1;
+    }
```

Comment:
```suggestion
    uint64_t val64 = value;
    if (IS_USED(MODULE_SYSTICK_PRESCALER)) {
        val64 *= prescaler;

        /* check if value fits with prescaler */
        if ((val64 >> 24) > ALARM_EXT_MAX) {
            return -1;
        }
    }
```

Reply:
If `prescaler` is one and is never written, the compiler will optimize it away.
I agree that we don't need to use a 64 bit variable here then, but the alarm can still be too far in the future.

Reply:
Looks like the compiler is even clever enough that a `unsigned int` multiplied by 1 will always fir into a `uint32_t`, so no 64 bit code is generated even with `uint64_t`.

https://godbolt.org/z/je1761